schema.sql

```diff
@@ -65,4 +65,24 @@ CREATE TABLE visits(
   vet_id INT REFERENCES vets(id),
   date DATE,
   PRIMARY KEY (id)
-);
\ No newline at end of file
+);
+
+ALTER TABLE owners ADD COLUMN email VARCHAR(120);
+
+ALTER TABLE visits
+RENAME COLUMN date TO date_of_visit;
+
+CREATE TABLE visits_animal_4 AS
+SELECT *
+FROM visits
+WHERE animal_id = 4;
+
+CREATE TABLE visits_vet_2 AS
+SELECT *
+FROM visits
+WHERE vet_id = 2;
+
+CREATE TABLE owner_18327 AS
+SELECT *
+FROM owners
```

Comment:
- [ ] Your execution plan `before` and `after` from the screenshot below did not improve instead it became worse. Please look into this. 
<br>
<img width="605" alt="image" src="https://github.com/deniodev/Vet-Clinic/assets/84156714/289f6f16-c0a4-4b8f-a812-ab41fed80fb8">

<br>

- [ ] Same thing with the second and third query - `SELECT * FROM visits where vet_id = 2;` and `SELECT \* FROM owners where email = 'owner_18327@mail.com';`

My recommendation is that you do NOT create new tables like this 👇🏾 

```sql
CREATE TABLE visits_animal_4 AS
SELECT *
FROM visits
WHERE animal_id = 4;

CREATE TABLE visits_vet_2 AS
SELECT *
FROM visits
WHERE vet_id = 2;

CREATE TABLE owner_18327 AS
SELECT *
FROM owners
```

Instead, you can add an index to the frequently visited columns in the table. Like this 👇🏾 , and this is done as the solution to the performance issue.

```sql
   CREATE INDEX ON visits (animal_id);
   CREATE INDEX ON visits (vet_id);
   CREATE INDEX ON owners (email);
```

Adding indexes to frequently visited columns in a table can significantly improve the performance of database queries. Indexes are data structures that help optimize the retrieval of data from database tables. They allow the database engine to quickly locate and retrieve specific rows based on the values in the indexed columns.

Here are a few reasons why adding indexes to frequently visited columns is beneficial:

1. Improved query performance: When a column is frequently used in search conditions or join operations, an index on that column allows the database engine to locate the desired rows more efficiently. It reduces the need for full table scans, resulting in faster query execution times.

2. Reduced data access time: Indexes organize the data in a way that reduces the amount of disk I/O required to access specific rows. By using index structures like B-trees or hash tables, the database engine can quickly navigate to the relevant data pages, minimizing the time needed to retrieve the desired information.

3. Enhanced concurrency: Indexes can improve concurrency by reducing the time locks are held on the data. With indexes, the database engine can access and modify specific rows more efficiently, reducing the duration of locks and allowing multiple users or processes to access the data concurrently.

4. Optimized sorting and grouping: Indexes can speed up sorting and grouping operations by providing a pre-sorted or pre-grouped data view. This can be particularly useful when working with large result sets or complex aggregations.

_It's worth noting that while indexes offer performance benefits, they also have some trade-offs. Indexes consume additional storage space, require maintenance when data is modified and can slow down write operations. Therefore, it's important to consider the specific characteristics of your application and choose indexing strategies wisely to strike a balance between improved query performance and the impact on other database operations._